Optional integration with c2isl

ezyang opened the pull request:
- We add c2isl as an optional dependency.  The recommended
  installation procedure is Conda; we'll pick up c2isl installs
  in $CONDA_PREFIX and use them if available.

- This commit set adds conversion functions between ATen tensors
  and dlpack.  The place they currently live is not long-term
  where we should put them, but as ATen does not presently have
  a dependency on the dlpack headers there isn't really anywhere
  we can put them at the moment.

- The code for dispatching to c2isl was heavily based off of
  the existing code for handling Caffe2.  We have a proof of
  concept operator using c2isl's MatMul implementation; gradient
  is NOT supported, and you must instantiate a new ISLMatMul class
  every time you want to run it on a different size.  You won't
  really get any performance increase this way, but this at least
  demonstrates it is possible to call to c2isl.

- In order to interface with c2isl, we have to allocate C-style
  DLTensor structures.  The memory management involved here is
  somewhat delicate and we have some custom deleters to handle
  the situation.

ezyang committed:
Optional integration with c2isl

- c2isl as added as an optional dependency.  The recommended
  installation procedure is Conda; we'll pick up c2isl installs
  in $CONDA_PREFIX and use them if available.

- This commit set adds conversion functions between ATen tensors
  and dlpack.  The place they currently live is not long-term
  where we should put them, but as ATen does not presently have
  a dependency on the dlpack headers there isn't really anywhere
  we can put them at the moment.

- The code for dispatching to c2isl was heavily based off of
  the existing code for handling Caffe2.  We have a proof of
  concept operator using c2isl's MatMul implementation; gradient
  is NOT supported, and you must instantiate a new ISLMatMul class
  every time you want to run it on a different size.  You won't
  really get any performance increase this way, but this at least
  demonstrates it is possible to call to c2isl.

- In order to interface with c2isl, we have to allocate C-style
  DLTensor structures.  The memory management involved here is
  somewhat delicate and we have some custom deleters to handle
  the situation.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on torch/csrc/autograd/c2isl_function.cpp:
```diff
@@ -0,0 +1,321 @@
+      t.code = DLDataTypeCode::kFloat;
+      t.bits = 16;
+      break;
+    case at::ScalarType::NumOptions:
```
`default` seems to be a better choice for raising an error. Same below

ezyang replied:
ATen's ScalarType is an enum class, so if we add default we lose exhaustiveness checking. I'll add a comment to this effect.

apaszke commented on torch/csrc/autograd/c2isl_function.cpp:
```diff
@@ -0,0 +1,321 @@
+}
+
+at::Type& toATenType(const DLContext& ctx, const DLDataType& t) {
+  if (t.lanes != 1) throw std::logic_error("ATen does not support lanes != 1");
```
It seems better to move this check to the overload above. This way it'd be called in both code paths + this one doesn't even use `t`

apaszke commented on torch/csrc/autograd/c2isl_function.cpp:
```diff
@@ -0,0 +1,321 @@
+  AutoGPU guard(input->data->getDevice());
+
+  if (!pImpl_) {
+    std::vector<const DLMetadata*> inMetas; // ownership managed by inTensorUPtrs
```
There's no such thing as `inTensorUPtrs`  did you mean `inMetaUPtrs`?

apaszke commented on torch/csrc/autograd/c2isl_function.cpp:
```diff
@@ -0,0 +1,321 @@
+    }
+    initTVM(inMetas);
+    pImpl_ = std::unique_ptr<c2isl::ISLTVMIROp>(
+              new c2isl::ISLTVMIROp(outputs, inputs, vars, ops)
```
Can you append underscores to member names? I've just spent some time looking where are they defined. I know we haven't done that in most places in autograd, however `vars` seems more like a local variable name to me

ezyang replied:
I gave them these names so we lined up with c2isl code better, but now that we have a Python interface this is less pressing, so I'll do this.

apaszke commented on torch/csrc/autograd/c2isl_function.cpp:
```diff
@@ -0,0 +1,321 @@
+}
+
+void ISLMatMul::initTVM(const std::vector<const DLMetadata*>& input_metas) {
+  // TODO: I feel dirty doing copy-assignment
```
this is move assigment

ezyang replied:
This code died.

apaszke commented on torch/csrc/autograd/c2isl_function.h:
```diff
@@ -0,0 +1,79 @@
+// Here's our strategy:
+//  - We new-allocate shape and strides for both DLTensor and DLMetadata,
+//    and free them in our deleters.
+//  - The deleter for DLTensor also holds a reference to at::Tensor, to keep
```
it holds an at::Tensor as a member, not as a reference

apaszke commented on torch/csrc/autograd/functions/init.cpp:
```diff
@@ -262,6 +274,11 @@ bool THPAutograd_initFunctions(PyObject* _unused)
+  #ifdef WITH_C2ISL
+  static PyTypeObject ISLMatMulClass;
+  addClass<ISLMatMul, ISLMatMulCtor>(module, ISLMatMulClass, "ISLMatMul");
+  #endif
```
This seems to be a debug code, and shouldn't be part of this PR

apaszke commented on torch/lib/ATen/templates/TensorDerived.cpp:
```diff
@@ -26,13 +26,18 @@ IntList ${Tensor}::sizes() {
+  int64_t d = ${THTensor}_nDimension(${state,}tensor);
+  if (d != 0)
+    return d;
+  return kUndefinedDimensions;
```
Why do we change the convention in here? All TH libs have 0-dim tensors

ezyang replied:
CC @killeent from who I got this patch.

apaszke commented on torch/csrc/autograd/c2isl_function.cpp:
```diff
@@ -0,0 +1,321 @@
+
+void ISLMatMul::initTVM(const std::vector<const DLMetadata*>& input_metas) {
+  // TODO: I feel dirty doing copy-assignment
+  vars = { ::tvm::Var("M"), ::tvm::Var("N"), ::tvm::Var("O") };
```
What's O used for?

ezyang replied:
This code died.

ezyang committed:
Add test for c2isl.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Support for specifying TVM IR in Python.

The big new user-facing module is torch.isl, which exposes
IslFunction, a function which can be subclassed with a TVM IR
description, and then on-the-fly compiles it into CUDA code.
See IslMatMult for an example of how to use it.

There is also a bugfix where THPUtils_unpackLong in python_numbers.h
is given a different name so it doesn't conflict with the similarly
named macro (causing very strange errors).  It will be worth working
out exactly what is going on here.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
More robust error checking.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Handle review comments.

- Remove test_c2isl.py from run_test.sh, since it doesn't work
  anyway.
- Rename torch.isl to torch._isl to indicate it is internal for now.
- Comment on case on at::ScalarType::NumOptions to shared
  toATenType function.
- Suffix all member variables with underscore.
- Prefix inputs/outputs/ops/vars with tvm (and suffix with underscore)
- Delete unnecessary gradient code

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on torch/_isl.py:
```diff
@@ -0,0 +1,179 @@
+class IslMatMul(IslFunction):
+    @staticmethod
+    def name():
+        return "matmul"
```
Is there any reason why we want to make it a method instead of a class attribute (`name = "matmul"`), or just use the class name?

ezyang replied:
Not really, I just picked something arbitrary. But one post facto justification is you might want to generate a name based on some parameters that are passed to the constructor; then you need a method.

apaszke replied:
Yes, except it can't be a static method that it is at the moment

ezyang committed:
Fixes for ATen.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Top-level gitmodules for gloo's submodules.

See https://stackoverflow.com/questions/45422693/how-to-use-git-repo-with-submodule-nested-in-subtree

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
